src/compiler/rust.rs

```diff
@@ -0,0 +1,761 @@
+const ALLOWED_EMIT: &'static [&'static str] = &["link", "dep-info"];
+
+/// Version number for cache key.
+const CACHE_VERSION: &'static [u8] = b"1";
+
+/// Parse the value passed to a commandline argument `arg`, either after
+/// an '=' in `arg` or in the next argument from `it`.
+fn arg_val<'a, 'b, T>(arg: &'a str, it: &'b mut T) -> Option<&'a str>
+    where T: Iterator<Item=&'a String>,
+{
+    if let Some(i) = arg.find('=') {
+        Some(&arg[i+1..])
+    } else {
+        it.next().map(|v| v.as_str())
+    }
+}
+
+/// Return true if `arg` is in the set of arguments `set`.
+fn arg_in(arg: &str, set: &HashSet<&str>) -> bool
+{
+    set.contains(arg) || set.iter().any(|a| arg.starts_with(a))
+}
+
+/// Calculate the SHA-1 digest of each file in `files` on background threads
+/// in `pool`.
+fn hash_all(files: Vec<String>, pool: &CpuPool) -> SFuture<Vec<String>>
+{
+    let start = Instant::now();
+    let count = files.len();
+    let pool = pool.clone();
+    Box::new(future::join_all(files.into_iter().map(move |f| sha1_digest(f, &pool)))
+             .map(move |hashes| {
+                 trace!("Hashed {} files in {}", count, fmt_duration_as_secs(&start.elapsed()));
+                 hashes
+             }))
+}
+
+/// Calculate SHA-1 digests for all source files listed in rustc's dep-info output.
+fn hash_source_files<T>(creator: &T, crate_name: &str, executable: &str, arguments: &[String], cwd: &str, pool: &CpuPool) -> SFuture<Vec<String>>
+    where T: CommandCreatorSync,
+{
+    let start = Instant::now();
+    // Get the full list of source files from rustc's dep-info.
+    let temp_dir = match TempDir::new("sccache") {
+        Ok(d) => d,
+        _ => return Box::new(future::err("Failed to create temp dir".into())),
+    };
+    let dep_file = temp_dir.path().join("deps.d");
+    let mut cmd = creator.clone().new_command_sync(executable);
+    cmd.args(&arguments)
+        .args(&["--emit", "dep-info"])
+        .arg("-o")
+        .arg(&dep_file)
+        .current_dir(cwd);
+    trace!("[{}]: get dep-info: {:?}", crate_name, cmd);
+    let dep_info = run_input_output(cmd, None);
+    // Parse the dep-info file, then hash the contents of those files.
+    let pool = pool.clone();
+    let cwd = cwd.to_owned();
+    let crate_name = crate_name.to_owned();
+    Box::new(dep_info.and_then(move |output| -> SFuture<_> {
+        if output.status.success() {
+            match parse_dep_file(&dep_file, &cwd) {
+                Ok(files) => {
+                    trace!("[{}]: got {} source files from dep-info in {}", crate_name,
+                           files.len(), fmt_duration_as_secs(&start.elapsed()));
+                    // Just to make sure we capture temp_dir.
+                    drop(temp_dir);
+                    hash_all(files, &pool)
+                }
+                Err(e) => return Box::new(future::err(e)),
+            }
+        } else {
+            Box::new(future::err("Failed run rustc --dep-info".into()))
```

Comment:
That's a good idea. We return the preprocessor stderr for C compilation when it fails, we should do something smarter here. The error propagation in this code isn't great in general, I should make sure we're providing enough information that users can attempt to diagnose failures.